File: implementations/typescript/packages/api-server/src/prompt-state.ts
// SPDX-License-Identifier: Apache-2.0
import {canonicalVersion,record} from "@psp-cdl/core";
export const PROMPT_REFRESH_PROFILE="PSP-PROMPT-REFRESH-0.1";
const numeric=(s:string)=>/^[0-9]+$/.test(s);
const compare=(a:string,b:string)=>a===b?0:a<b?-1:1;
const numberCompare=(a:string,b:string)=>a.length===b.length?compare(a,b):a.length<b.length?-1:1;
/** SemVer precedence, without converting arbitrarily long version numbers to floats. */
export function comparePromptVersions(a:string,b:string):number {
  const parts=(s:string)=>{const v=canonicalVersion(s).split('+')[0]!,dash=v.indexOf('-');return {core:(dash<0?v:v.slice(0,dash)).split('.'),pre:dash<0?[]:v.slice(dash+1).split('.')};};
  const x=parts(a),y=parts(b);
  for(let i=0;i<3;i++){const n=numberCompare(x.core[i]!,y.core[i]!);if(n)return n;}
  if(!x.pre.length||!y.pre.length)return x.pre.length===y.pre.length?0:x.pre.length?-1:1;
  for(let i=0;i<Math.min(x.pre.length,y.pre.length);i++){
    const u=x.pre[i]!,v=y.pre[i]!,n=numeric(u)&&numeric(v)?numberCompare(u,v):numeric(u)!==numeric(v)?numeric(u)?-1:1:compare(u,v);if(n)return n;
  }
  return Math.sign(x.pre.length-y.pre.length);
}
export function validPromptState(s:unknown):boolean {
  if(!record(s)||Object.keys(s).sort().join(',')!=="digest,expires,grace,interval,policies,refreshCount,timestamp,turnCount,version")return false;
  try{if(canonicalVersion(s.version as string)!==s.version)return false;}catch{return false;}
  if(typeof s.digest!=="string"||s.digest.length!==64||!/^[a-f0-9]{64}$/.test(s.digest))return false;
  for(const k of ["expires","grace","interval","refreshCount","timestamp","turnCount"])if(typeof s[k]!=="number"||!Number.isSafeInteger(s[k])||(s[k] as number)<0)return false;
  return (s.expires as number)>(s.timestamp as number)&&Array.isArray(s.policies)&&s.policies.length>0&&s.policies.length<=2&&
    s.policies.every((p,i,list)=>["expiration","interval"].includes(p)&&list.indexOf(p)===i)&&
    s.policies.includes("interval")===(s.interval as number>0);
}

File: implementations/python/packages/api-server/src/psp_cdl_api_server/prompt_state.py
# SPDX-License-Identifier: Apache-2.0
import re
from psp_cdl_core import canonical_version
PROMPT_REFRESH_PROFILE="PSP-PROMPT-REFRESH-0.1"

def compare_prompt_versions(a,b):
    def parts(s):
        core,sep,pre=canonical_version(s).split("+",1)[0].partition("-")
        return core.split("."),pre.split(".") if sep else []
    def cmp(a,b): return (a>b)-(a<b)
    def num(a,b): return cmp(len(a),len(b)) or cmp(a,b)
    x,xpre=parts(a);y,ypre=parts(b)
    for u,v in zip(x,y):
        n=num(u,v)
        if n:return n
    if not xpre or not ypre:return cmp(bool(ypre),bool(xpre))
    for u,v in zip(xpre,ypre):
        n=num(u,v) if u.isascii() and u.isdigit() and v.isascii() and v.isdigit() else cmp(not u.isdigit(),not v.isdigit()) if u.isdigit()!=v.isdigit() else cmp(u,v)
        if n:return n
    return cmp(len(xpre),len(ypre))

def valid_prompt_state(s):
    if type(s) is not dict or set(s)!={"digest","expires","grace","interval","policies","refreshCount","timestamp","turnCount","version"}:return False
    try:
        if canonical_version(s["version"])!=s["version"]:return False
    except Exception:return False
    if type(s["digest"]) is not str or not re.fullmatch(r"[a-f0-9]{64}",s["digest"]):return False
    for k in ("expires","grace","interval","refreshCount","timestamp","turnCount"):
        n=s[k]
        if type(n) not in (int,float) or not 0<=n<=9007199254740991 or int(n)!=n:return False
    p=s["policies"]
    return s["expires"]>s["timestamp"] and type(p) is list and 0<len(p)<=2 and all(type(v) is str and v in ("expiration","interval") and p.index(v)==i for i,v in enumerate(p)) and (("interval" in p)==(s["interval"]>0))
